### PR TITLE
Build can store docker image to private registry when cluster uses official raw docker

### PR DESCRIPTION
**Pull request by shy-Xu**
### Describe what this PR does / why we need it
this PR is aimed to solve the issue #865

### CODE REVIEW

**Comment by shy-Xu** [comment]
After discusstion with @kakaZhou719 ,I design a way to solve issue #865.Now,I add a flag to identify whether the cluster use raw docker.I also complete the interface definations and functions call.I will finish the implement of these functions later.

**Comment by shy-Xu** [comment]
I have completed the code development, but the check is failed for unknow reason

File: client/docker/images.go [hunk 1]
```diff
@@ -29,32 +29,92 @@ import (
...
+		if strings.HasPrefix(image, "#") {
+			continue
+		}
+		if err := d.ImagePull(strings.TrimSpace(image)); err != nil {
+			return fmt.Errorf("image %s pull failed: %v", image, err)
+		}
+	}
+	return nil
+}
+
```
Comment (shy-Xu) [inline client/docker/images.go:106]: I comment out these two functions because I find they are not used

**Comment by shy-Xu** [comment]
Now users need not input a flag to identify whether it's sealer docker or raw docker

File: build/buildkit/buildimage/build_image.go [hunk 2]
```diff
@@ -62,6 +64,16 @@ func (b BuildImage) ExecBuild(ctx Context) error {
 	} else {
 		execCtx = buildinstruction.NewExecContextWithoutCache(ctx.BuildType, ctx.BuildContext, b.LayerStore)
 	}
+	//judge the docker engine version
+	dockerServerVersion, err := b.DockerClient.GetServerVersion()
```
Comment (fanux) [inline build/buildkit/buildimage/build_image.go:68]: Check the version in `ImagesCacheToRegistry`, so you don't need the bool argument

File: client/docker/images.go [hunk 3]
```diff
@@ -29,32 +29,96 @@ import (
...
+	err = d.cli.ImageTag(d.ctx, named.String(), TagImageName)
+	if err != nil {
+		return err
+	}
+	out, err = d.cli.ImagePush(d.ctx, TagImageName, opts)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		_ = out.Close()
```
Comment (fanux) [inline client/docker/images.go:86]: should not ignore errors

File: client/docker/images.go [hunk 4]
```diff
@@ -29,32 +29,96 @@ import (
...
 	if err != nil {
-		logger.Error(fmt.Sprintf("Read image list failed: %v", err))
+		return fmt.Errorf("failed to parse canonical image name %s : %v", image, err)
 	}
-	return d.ImagesPull(data)
-}
+	TagImageName = strings.Replace(named.String(), named.String()[0:strings.Index(named.String(), `/`)], registryName, 1)
+	opts := types.ImagePushOptions(GetCanonicalImagePullOptions(named.String()))
+	if opts.RegistryAuth == "" {
+		opts.RegistryAuth = "sea.hub" //must not empty
```
Comment (fanux) [inline client/docker/images.go:75]: Use const

File: client/docker/images.go [hunk 5]
```diff
@@ -29,32 +29,96 @@ import (
 	dockerjsonmessage "github.com/docker/docker/pkg/jsonmessage"
 )
 
-func (d Docker) ImagesPull(images []string) error {
+func (d Docker) ImagesCacheToRegistry(images []string) error {
+	version, err := d.cli.ServerVersion(d.ctx)
+	if err != nil {
+		return err
+	}
+	if strings.Contains(version.Version, `sealer`) {
```
Comment (fanux) [inline client/docker/images.go:37]: const

File: client/docker/images.go [hunk 6]
```diff
@@ -29,32 +29,96 @@ import (
...
+		TagImageName string
+	)
+	named, err = GetCanonicalImageName(image)
 	if err != nil {
-		logger.Error(fmt.Sprintf("Read image list failed: %v", err))
+		return fmt.Errorf("failed to parse canonical image name %s : %v", image, err)
 	}
-	return d.ImagesPull(data)
-}
+	TagImageName = strings.Replace(named.String(), named.String()[0:strings.Index(named.String(), `/`)], registryName, 1)
```
Comment (fanux) [inline client/docker/images.go:72]: How about the image name don't contains `/` ?

Comment (shy-Xu) [reply client/docker/images.go:72]: named.String() returns a image name which has been handled. For example, if the input image name is "ubuntu", named.String() returns docker.io/library/ubuntu:latest

**Review by fanux — APPROVED** [review]

File: client/docker/images.go [hunk 7]
```diff
@@ -29,6 +29,27 @@ import (
...
 )
 
+const (
+	sealerDockerString = `sealer`
+	registryAuthString = `sea.hub`
+	registryName       = `sea.hub:5000`
+)
+
+func (d Docker) ImagesCacheToRegistry(images []string) error {
+	version, err := d.cli.ServerVersion(d.ctx)
```
Comment (kakaZhou719) [inline client/docker/images.go:39]: whether is sealer-docker or not, we can move out this as an utils flag. And in some scenarios we don’t even have docker, we can try to use cmd "docker version | grep sealer".

**Pull request closed without merging**